woocommerce/woocommerce-ios
# Support `UIImage` in product image upload

Prerequisite for HACK Week pe5pgL-3Bl-p2#comment-2957


## Why


In order to implement image processing like background removal for product images, the app needs to support image type `UIImage` in addition to `PHAsset` from the camera or device photo library. Unlike `PHAsset` which contains camera metadata, the filename and alt text need to be provided manually for a `UIImage` after image processing. This requires some changes in how we export and upload media. The changes in this PR are not expected to affect the app behavior, and the `UIImage` use case will be in a separate PR.

## How

First, `ProductImageStatus.uploading` enum case's associated type has been changed from `PHAsset` to a new enum type `ProductImageAssetType` which now can be either `PHAsset` or `UIImage` (plus filename and alt text). This change results in changes everywhere `ProductImageStatus.uploading` is used to also accommodate `UIImage`.

Then, the core changes are in the Yosemite layer:

- `MediaUIImageExporter` was created for exporting a `UIImage` to `UploadableMedia`
- In `MediaExportService`, it uses the new `MediaUIImageExporter` when the asset is `UIImage`
- In `MediaImageExporter` where both `PHAsset` and `UIImage` exporters call to export from image `Data` to `UploadableMedia`, the alt text is DI'ed and set in the output `UploadableMedia`
- `UIImage` now conforms to `ExportableAsset`, and `MediaAction` now includes `filename` and `altText` for `UIImage` or if we want to support metadata editing for all image types

Networking layer:

In `MediaRemote.uploadMediaToWordPressSite`, the interface was updated to only take in one media item because the WP API doesn't support multiple items. The return type is also already returning one media item. In both `MediaRemote.uploadMedia` functions, the alt text is now included in the network request. The implementation differs when using the WPCOM and site API.

## Testing instructions


There's no use case for the `UIImage` asset type right now, the background removal feature is only available for iOS 17+ so we need to wait for Xcode 15 to be adopted in CI first. For confidence check to make sure the changes don't affect the current image upload:

- Log in to a store if needed
- Go to the Products tab
- Tap on an editable product
- Tap on the images header
- Tap `Add Photos`
- Select either `Choose from device` and pick some photos --> after confirming the selection, the images should be uploaded to the site's media library in core successfully and have a filename that's not UUID (unless that's the actual filename of the image)
- Tap `Save` --> after the images are uploaded, the product should include the uploaded images

---

- [x] @jaclync tests the camera flow
- [x] @jaclync tests the image upload flow in a JCP site 

---
